File: dart/lib/src/ast.dart
/// AST types for hron schedule expressions.
///
/// These types represent the parsed structure of hron expressions and are
/// exposed for advanced use cases like custom schedule analysis.
library;

/// Days of the week.
enum Weekday {
  monday,
  tuesday,
  wednesday,
  thursday,
  friday,
  saturday,
  sunday;

  int get number => index + 1; // 1=Monday ... 7=Sunday

  int get cronDow {
    const map = [1, 2, 3, 4, 5, 6, 0]; // mon=1..sat=6, sun=0
    return map[index];
  }

  static Weekday? tryParse(String s) => _weekdayMap[s.toLowerCase()];

  static Weekday fromNumber(int n) => Weekday.values[n - 1];

  static Weekday fromCronDow(int n) {
    const map = {
      0: Weekday.sunday,
      1: Weekday.monday,
      2: Weekday.tuesday,
      3: Weekday.wednesday,
      4: Weekday.thursday,
      5: Weekday.friday,
      6: Weekday.saturday,
      7: Weekday.sunday,
    };
    return map[n]!;
  }
}

const _weekdayMap = {
  'monday': Weekday.monday,
  'mon': Weekday.monday,
  'tuesday': Weekday.tuesday,
  'tue': Weekday.tuesday,
  'wednesday': Weekday.wednesday,
  'wed': Weekday.wednesday,
  'thursday': Weekday.thursday,
  'thu': Weekday.thursday,
  'friday': Weekday.friday,
  'fri': Weekday.friday,
  'saturday': Weekday.saturday,
  'sat': Weekday.saturday,
  'sunday': Weekday.sunday,
  'sun': Weekday.sunday,
};

/// Month names for date specifications.
enum MonthName {
  jan,
  feb,
  mar,
  apr,
  may,
  jun,
  jul,
  aug,
  sep,
  oct,
  nov,
  dec;

  int get number => index + 1;

  static MonthName? tryParse(String s) => _monthMap[s.toLowerCase()];

  static MonthName fromNumber(int n) => MonthName.values[n - 1];
}

const _monthMap = {
  'january': MonthName.jan,
  'jan': MonthName.jan,
  'february': MonthName.feb,
  'feb': MonthName.feb,
  'march': MonthName.mar,
  'mar': MonthName.mar,
  'april': MonthName.apr,
  'apr': MonthName.apr,
  'may': MonthName.may,
  'june': MonthName.jun,
  'jun': MonthName.jun,
  'july': MonthName.jul,
  'jul': MonthName.jul,
  'august': MonthName.aug,
  'aug': MonthName.aug,
  'september': MonthName.sep,
  'sep': MonthName.sep,
  'october': MonthName.oct,
  'oct': MonthName.oct,
  'november': MonthName.nov,
  'nov': MonthName.nov,
  'december': MonthName.dec,
  'dec': MonthName.dec,
};

/// Time interval units for repeat expressions.
enum IntervalUnit {
  /// Minutes.
  min,

  /// Hours.
  hours,
}

/// Ordinal positions for expressions like "first Monday of every month".
enum OrdinalPosition {
  first,
  second,
  third,
  fourth,
  fifth,
  last;

  int get toN {
    const map = {
      OrdinalPosition.first: 1,
      OrdinalPosition.second: 2,
      OrdinalPosition.third: 3,
      OrdinalPosition.fourth: 4,
      OrdinalPosition.fifth: 5,
    };
    return map[this]!;
  }
}

/// A time of day (hour and minute) without timezone.
class TimeOfDay {
  final int hour;
  final int minute;

  const TimeOfDay(this.hour, this.minute);

  @override
  bool operator ==(Object other) =>
      other is TimeOfDay && other.hour == hour && other.minute == minute;

  @override
  int get hashCode => Object.hash(hour, minute);

  @override
  String toString() =>
      '${hour.toString().padLeft(2, '0')}:${minute.toString().padLeft(2, '0')}';
}

/// Filter for which days a schedule applies to.
sealed class DayFilter {}

/// Matches every day of the week.
class EveryDay extends DayFilter {}

/// Matches Monday through Friday.
class WeekdayFilter extends DayFilter {}

/// Matches Saturday and Sunday.
class WeekendFilter extends DayFilter {}

/// Matches specific days of the week.
class SpecificDays extends DayFilter {
  final List<Weekday> days;
  SpecificDays(this.days);
}

/// Specification for days within a month.
sealed class DayOfMonthSpec {}

class SingleDay extends DayOfMonthSpec {
  final int day;
  SingleDay(this.day);
}

class DayRange extends DayOfMonthSpec {
  final int start;
  final int end;
  DayRange(this.start, this.end);
}

/// Target specification for monthly schedules.
sealed class MonthTarget {}

class DaysTarget extends MonthTarget {
  final List<DayOfMonthSpec> specs;
  DaysTarget(this.specs);
}

class LastDayTarget extends MonthTarget {}

class LastWeekdayTarget extends MonthTarget {}

/// Target specification for yearly schedules.
sealed class YearTarget {}

class DateTarget extends YearTarget {
  final MonthName month;
  final int day;
  DateTarget(this.month, this.day);
}

class OrdinalWeekdayTarget extends YearTarget {
  final OrdinalPosition ordinal;
  final Weekday weekday;
  final MonthName month;
  OrdinalWeekdayTarget(this.ordinal, this.weekday, this.month);
}

class DayOfMonthTarget extends YearTarget {
  final int day;
  final MonthName month;
  DayOfMonthTarget(this.day, this.month);
}

class LastWeekdayYearTarget extends YearTarget {
  final MonthName month;
  LastWeekdayYearTarget(this.month);
}

/// A specific date (named like "Jan 1" or ISO like "2024-01-01").
sealed class DateSpec {}

class NamedDate extends DateSpec {
  final MonthName month;
  final int day;
  NamedDate(this.month, this.day);
}

class IsoDate extends DateSpec {
  final String date;
  IsoDate(this.date);
}

/// A date to exclude from a schedule (used in `except` clauses).
sealed class ExceptionSpec {}

class NamedException extends ExceptionSpec {
  final MonthName month;
  final int day;
  NamedException(this.month, this.day);
}

class IsoException extends ExceptionSpec {
  final String date;
  IsoException(this.date);
}

/// End date for a schedule (used in `until` clauses).
sealed class UntilSpec {}

class IsoUntil extends UntilSpec {
  final String date;
  IsoUntil(this.date);
}

class NamedUntil extends UntilSpec {
  final MonthName month;
  final int day;
  NamedUntil(this.month, this.day);
}

/// A schedule expression representing when events occur.
///
/// This is the main AST type representing different schedule patterns:
/// - [IntervalRepeat]: "every 30 min from 9am to 5pm"
/// - [DayRepeat]: "every day at 9am"
/// - [WeekRepeat]: "every week on monday at 9am"
/// - [MonthRepeat]: "on the 1st of every month at 9am"
/// - [OrdinalRepeat]: "on the first monday of every month at 9am"
/// - [SingleDate]: "on Jan 1 at 12:00"
/// - [YearRepeat]: "every year on Jan 1 at 12:00"
sealed class ScheduleExpr {}

/// Schedule repeating at a time interval (e.g., "every 30 min from 9am to 5pm").
class IntervalRepeat extends ScheduleExpr {
  final int interval;
  final IntervalUnit unit;
  final TimeOfDay from;
  final TimeOfDay to;
  final DayFilter? dayFilter;
  IntervalRepeat(this.interval, this.unit, this.from, this.to, this.dayFilter);
}

/// Schedule repeating daily (e.g., "every day at 9am").
class DayRepeat extends ScheduleExpr {
  final int interval;
  final DayFilter days;
  final List<TimeOfDay> times;
  DayRepeat(this.interval, this.days, this.times);
}

/// Schedule repeating weekly (e.g., "every week on monday at 9am").
class WeekRepeat extends ScheduleExpr {
  final int interval;
  final List<Weekday> days;
  final List<TimeOfDay> times;
  WeekRepeat(this.interval, this.days, this.times);
}

/// Schedule repeating monthly (e.g., "on the 1st of every month at 9am").
class MonthRepeat extends ScheduleExpr {
  final int interval;
  final MonthTarget target;
  final List<TimeOfDay> times;
  MonthRepeat(this.interval, this.target, this.times);
}

/// Schedule repeating on ordinal weekdays (e.g., "first monday of every month").
class OrdinalRepeat extends ScheduleExpr {
  final int interval;
  final OrdinalPosition ordinal;
  final Weekday day;
  final List<TimeOfDay> times;
  OrdinalRepeat(this.interval, this.ordinal, this.day, this.times);
}

/// A one-time schedule on a specific date (e.g., "on Jan 1 at 12:00").
class SingleDate extends ScheduleExpr {
  final DateSpec date;
  final List<TimeOfDay> times;
  SingleDate(this.date, this.times);
}

/// Schedule repeating yearly (e.g., "every year on Jan 1 at 12:00").
class YearRepeat extends ScheduleExpr {
  final int interval;
  final YearTarget target;
  final List<TimeOfDay> times;
  YearRepeat(this.interval, this.target, this.times);
}

/// The complete parsed schedule with expression and modifiers.
///
/// Contains the main [expr] plus optional clauses like timezone,
/// exceptions, end date, anchor, and month restrictions.
class ScheduleData {
  final ScheduleExpr expr;
  String? timezone;
  List<ExceptionSpec> except;
  UntilSpec? until;
  String? anchor;
  List<MonthName> during;

  ScheduleData(this.expr) : except = [], during = [];
}

// --- Helper functions ---

List<int> expandDaySpec(DayOfMonthSpec spec) {
  if (spec is SingleDay) return [spec.day];
  final range = spec as DayRange;
  return [for (var d = range.start; d <= range.end; d++) d];
}

List<int> expandMonthTarget(MonthTarget target) {
  if (target is DaysTarget) {
    return target.specs.expand(expandDaySpec).toList();
  }
  return [];
}

String ordinalSuffix(int n) {
  final mod100 = n % 100;
  if (mod100 >= 11 && mod100 <= 13) return 'th';
  switch (n % 10) {
    case 1:
      return 'st';
    case 2:
      return 'nd';
    case 3:
      return 'rd';
    default:
      return 'th';
  }
}

File: rust/hron/src/ast.rs
#[cfg(feature = "serde")]
use serde::{Deserialize, Deserializer, Serialize, Serializer};

/// A parsed hron schedule: expression + optional modifiers.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Schedule {
    pub expr: ScheduleExpr,
    pub timezone: Option<String>,
    pub except: Vec<Exception>,
    pub until: Option<UntilSpec>,
    pub anchor: Option<jiff::civil::Date>,
    pub during: Vec<MonthName>,
}

impl Schedule {
    /// Create a Schedule from just an expression (no modifiers).
    pub fn new(expr: ScheduleExpr) -> Self {
        Self {
            expr,
            timezone: None,
            except: Vec::new(),
            until: None,
            anchor: None,
            during: Vec::new(),
        }
    }
}

/// The core schedule expression (what repeats).
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ScheduleExpr {
    /// `every 30 min from 09:00 to 17:00 [on weekdays]`
    IntervalRepeat {
        interval: u32,
        unit: IntervalUnit,
        from: TimeOfDay,
        to: TimeOfDay,
        day_filter: Option<DayFilter>,
    },
    /// `every day at 09:00`, `every 2 days at 09:00`
    DayRepeat {
        interval: u32,
        days: DayFilter,
        times: Vec<TimeOfDay>,
    },
    /// `every 2 weeks on monday at 09:00`
    WeekRepeat {
        interval: u32,
        days: Vec<Weekday>,
        times: Vec<TimeOfDay>,
    },
    /// `every month on the 1st at 09:00`, `every 2 months on the 1st at 09:00`
    MonthRepeat {
        interval: u32,
        target: MonthTarget,
        times: Vec<TimeOfDay>,
    },
    /// `first monday of every month at 10:00`, `first monday of every 2 months at 10:00`
    OrdinalRepeat {
        interval: u32,
        ordinal: OrdinalPosition,
        day: Weekday,
        times: Vec<TimeOfDay>,
    },
    /// `on feb 14 at 9:00, 17:00`
    SingleDate {
        date: DateSpec,
        times: Vec<TimeOfDay>,
    },
    /// `every year on dec 25 at 00:00`, `every 2 years on dec 25 at 00:00`
    YearRepeat {
        interval: u32,
        target: YearTarget,
        times: Vec<TimeOfDay>,
    },
}

/// Exception date for `except` clause.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "snake_case"))]
pub enum Exception {
    /// Recurring named date: `dec 25` matches every year.
    Named { month: MonthName, day: u8 },
    /// One-off ISO date: `2026-12-25`.
    Iso(String),
}

/// Until spec for `until` clause.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "snake_case"))]
pub enum UntilSpec {
    /// ISO date: `2026-12-31`.
    Iso(String),
    /// Named date: `dec 31` — resolves to next occurrence from current year.
    Named { month: MonthName, day: u8 },
}

/// Year target for yearly expressions.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "snake_case"))]
pub enum YearTarget {
    /// `on dec 25` — specific month and day.
    Date { month: MonthName, day: u8 },
    /// `on the first monday of march` — ordinal weekday of a month.
    OrdinalWeekday {
        ordinal: OrdinalPosition,
        weekday: Weekday,
        month: MonthName,
    },
    /// `on the 15th of march` — day of a month.
    DayOfMonth { day: u8, month: MonthName },
    /// `on the last weekday of december` — last weekday of a month.
    LastWeekday { month: MonthName },
}

/// Time of day (hours and minutes).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TimeOfDay {
    pub hour: u8,
    pub minute: u8,
}

#[cfg(feature = "serde")]
impl Serialize for TimeOfDay {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        serializer.serialize_str(&format!("{:02}:{:02}", self.hour, self.minute))
    }
}

#[cfg(feature = "serde")]
impl<'de> Deserialize<'de> for TimeOfDay {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let s = String::deserialize(deserializer)?;
        let parts: Vec<&str> = s.split(':').collect();
        if parts.len() != 2 {
            return Err(serde::de::Error::custom("expected HH:MM"));
        }
        let hour = parts[0]
            .parse()
            .map_err(|_| serde::de::Error::custom("invalid hour"))?;
        let minute = parts[1]
            .parse()
            .map_err(|_| serde::de::Error::custom("invalid minute"))?;
        Ok(TimeOfDay { hour, minute })
    }
}

/// Day filter for day-repeat and interval expressions.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "lowercase"))]
pub enum DayFilter {
    Every,
    Weekday,
    Weekend,
    Days(Vec<Weekday>),
}

/// Weekday with custom serde (lowercase string).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum Weekday {
    Monday,
    Tuesday,
    Wednesday,
    Thursday,
    Friday,
    Saturday,
    Sunday,
}

impl Weekday {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Monday => "monday",
            Self::Tuesday => "tuesday",
            Self::Wednesday => "wednesday",
            Self::Thursday => "thursday",
            Self::Friday => "friday",
            Self::Saturday => "saturday",
            Self::Sunday => "sunday",
        }
    }

    pub fn short(self) -> &'static str {
        match self {
            Self::Monday => "mon",
            Self::Tuesday => "tue",
            Self::Wednesday => "wed",
            Self::Thursday => "thu",
            Self::Friday => "fri",
            Self::Saturday => "sat",
            Self::Sunday => "sun",
        }
    }

    pub fn to_jiff(self) -> jiff::civil::Weekday {
        match self {
            Self::Monday => jiff::civil::Weekday::Monday,
            Self::Tuesday => jiff::civil::Weekday::Tuesday,
            Self::Wednesday => jiff::civil::Weekday::Wednesday,
            Self::Thursday => jiff::civil::Weekday::Thursday,
            Self::Friday => jiff::civil::Weekday::Friday,
            Self::Saturday => jiff::civil::Weekday::Saturday,
            Self::Sunday => jiff::civil::Weekday::Sunday,
        }
    }

    pub fn from_jiff(wd: jiff::civil::Weekday) -> Self {
        match wd {
            jiff::civil::Weekday::Monday => Self::Monday,
            jiff::civil::Weekday::Tuesday => Self::Tuesday,
            jiff::civil::Weekday::Wednesday => Self::Wednesday,
            jiff::civil::Weekday::Thursday => Self::Thursday,
            jiff::civil::Weekday::Friday => Self::Friday,
            jiff::civil::Weekday::Saturday => Self::Saturday,
            jiff::civil::Weekday::Sunday => Self::Sunday,
        }
    }

    /// ISO 8601 day number: Monday=1, Sunday=7.
    pub fn number(self) -> u8 {
        match self {
            Self::Monday => 1,
            Self::Tuesday => 2,
            Self::Wednesday => 3,
            Self::Thursday => 4,
            Self::Friday => 5,
            Self::Saturday => 6,
            Self::Sunday => 7,
        }
    }

    pub fn from_number(n: u8) -> Option<Self> {
        match n {
            1 => Some(Self::Monday),
            2 => Some(Self::Tuesday),
            3 => Some(Self::Wednesday),
            4 => Some(Self::Thursday),
            5 => Some(Self::Friday),
            6 => Some(Self::Saturday),
            7 => Some(Self::Sunday),
            _ => None,
        }
    }

    pub fn all_weekdays() -> Vec<Self> {
        vec![
            Self::Monday,
            Self::Tuesday,
            Self::Wednesday,
            Self::Thursday,
            Self::Friday,
        ]
    }

    pub fn all_weekend() -> Vec<Self> {
        vec![Self::Saturday, Self::Sunday]
    }
}

#[cfg(feature = "serde")]
impl Serialize for Weekday {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        serializer.serialize_str(self.as_str())
    }
}

#[cfg(feature = "serde")]
impl<'de> Deserialize<'de> for Weekday {
    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let s = String::deserialize(deserializer)?;
        parse_weekday(&s).ok_or_else(|| serde::de::Error::custom(format!("unknown weekday: {s}")))
    }
}

pub fn parse_weekday(s: &str) -> Option<Weekday> {
    match s.to_lowercase().as_str() {
        "monday" | "mon" => Some(Weekday::Monday),
        "tuesday" | "tue" => Some(Weekday::Tuesday),
        "wednesday" | "wed" => Some(Weekday::Wednesday),
        "thursday" | "thu" => Some(Weekday::Thursday),
        "friday" | "fri" => Some(Weekday::Friday),
        "saturday" | "sat" => Some(Weekday::Saturday),
        "sunday" | "sun" => Some(Weekday::Sunday),
        _ => None,
    }
}

/// A single day or range of days in a monthly target.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "snake_case"))]
pub enum DayOfMonthSpec {
    Single(u8),
    Range(u8, u8),
}

impl DayOfMonthSpec {
    /// Expand into individual day numbers.
    pub fn expand(&self) -> Vec<u8> {
        match self {
            DayOfMonthSpec::Single(d) => vec![*d],
            DayOfMonthSpec::Range(start, end) => (*start..=*end).collect(),
        }
    }
}

/// Direction for nearest weekday (hron extension beyond cron W).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "snake_case"))]
pub enum NearestDirection {
    /// Always prefer following weekday (can cross to next month).
    Next,
    /// Always prefer preceding weekday (can cross to prev month).
    Previous,
}

/// Month target for month-repeat expressions.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "snake_case"))]
pub enum MonthTarget {
    Days(Vec<DayOfMonthSpec>),
    LastDay,
    LastWeekday,
    /// Nearest weekday to a given day of month.
    /// Standard (None): never crosses month boundary (cron W compatibility).
    /// Directional (Some): can cross month boundary.
    NearestWeekday {
        day: u8,
        direction: Option<NearestDirection>,
    },
}

impl MonthTarget {
    /// Expand all day specs into individual day numbers.
    pub fn expand_days(&self) -> Vec<u8> {
        match self {
            MonthTarget::Days(specs) => specs.iter().flat_map(|s| s.expand()).collect(),
            _ => vec![],
        }
    }
}

/// Ordinal position (first through fifth, or last).
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "lowercase"))]
pub enum OrdinalPosition {
    First,
    Second,
    Third,
    Fourth,
    Fifth,
    Last,
}

impl OrdinalPosition {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::First => "first",
            Self::Second => "second",
            Self::Third => "third",
            Self::Fourth => "fourth",
            Self::Fifth => "fifth",
            Self::Last => "last",
        }
    }
}

/// Date specification for single-date expressions.
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "snake_case"))]
pub enum DateSpec {
    Named { month: MonthName, day: u8 },
    Iso(String),
}

/// Month name.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "lowercase"))]
pub enum MonthName {
    January,
    February,
    March,
    April,
    May,
    June,
    July,
    August,
    September,
    October,
    November,
    December,
}

impl MonthName {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::January => "jan",
            Self::February => "feb",
            Self::March => "mar",
            Self::April => "apr",
            Self::May => "may",
            Self::June => "jun",
            Self::July => "jul",
            Self::August => "aug",
            Self::September => "sep",
            Self::October => "oct",
            Self::November => "nov",
            Self::December => "dec",
        }
    }

    pub fn number(self) -> u8 {
        match self {
            Self::January => 1,
            Self::February => 2,
            Self::March => 3,
            Self::April => 4,
            Self::May => 5,
            Self::June => 6,
            Self::July => 7,
            Self::August => 8,
            Self::September => 9,
            Self::October => 10,
            Self::November => 11,
            Self::December => 12,
        }
    }
}

pub fn parse_month_name(s: &str) -> Option<MonthName> {
    match s.to_lowercase().as_str() {
        "january" | "jan" => Some(MonthName::January),
        "february" | "feb" => Some(MonthName::February),
        "march" | "mar" => Some(MonthName::March),
        "april" | "apr" => Some(MonthName::April),
        "may" => Some(MonthName::May),
        "june" | "jun" => Some(MonthName::June),
        "july" | "jul" => Some(MonthName::July),
        "august" | "aug" => Some(MonthName::August),
        "september" | "sep" => Some(MonthName::September),
        "october" | "oct" => Some(MonthName::October),
        "november" | "nov" => Some(MonthName::November),
        "december" | "dec" => Some(MonthName::December),
        _ => None,
    }
}

/// Interval unit.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
#[cfg_attr(feature = "serde", serde(rename_all = "lowercase"))]
pub enum IntervalUnit {
    Minutes,
    Hours,
}

impl IntervalUnit {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::Minutes => "min",
            Self::Hours => "hours",
        }
    }

    pub fn singular(self) -> &'static str {
        match self {
            Self::Minutes => "minute",
            Self::Hours => "hour",
        }
    }
}
